backends/tc/ir_builder.h

```diff
@@ -0,0 +1,195 @@
+  // variables encountered to the correct declaration.
+  const P4::ReferenceMap& ref_map_;
+  // Mapping of P4 IR nodes to their types from the P4C front-end
+  const P4::TypeMap& type_map_;
+  // The generated TCAM program
+  TCAMProgram tcam_program_;
+  // The header parameter used by the P4 parser, we use this to find the
+  // extracted headers.
+  absl::optional<IR::ID> header_param_;
+
+  // Methods supported by this backend, mapped to their implementations.
+  const absl::flat_hash_map<
+      absl::string_view,
+      std::function<bool(IRBuilder&, const IR::MethodCallStatement*)>>
+      kSupportedMethods = {
+          {"extract", [](IRBuilder& ir_builder,
+                         const IR::MethodCallStatement* method_call) {
+             return ir_builder.ExtractImpl(method_call);
+           }}};
+
+  // Name of packet.lookahead, this method is kept separately from other methods
+  // we support because it occurs in a different context (right-hand side of
+  // assignments) from other methods (other methods appear as top-level
+  // statements). So, its functionality requires additional context that is not
+  // supplied to the method implementations.
+  const IR::ID kLookahead = "lookahead";
+
+  // Code generation information per state, this is used for keeping track of
+  // available fields and how many bits the parser has advanced in the middle of
+  // execution.
+  absl::flat_hash_map<std::string, CodeGenInfo> code_gen_info_;
+
+  // Resolve an expression that should accesses a field of the header parameter
+  // in the parser. It returns a vector representing the sequence of field
+  // accesses. For example, on the input `hdr.foo.bar` it returns ["foo", "bar"]
+  // if hdr is the header parameter of the parser. It returns nullopt if the
+  // expression is not a field access, or if it is accessing the fields of a
+  // variable other than header_param_.
+  absl::optional<std::vector<IR::ID>> ResolveHeaderExpression(
+      const IR::Expression* expr) const;
+
+  // Get ranges from the packet buffer relative to the cursor at the beginning
+  // of current state. This is used for concatenating values and emitting
+  // set-key instructions.
+  std::vector<util::Range> GetRangesRelativeToCursor(
+      const std::string& current_state,
+      const IR::ListExpression* select_guard) const;
+
+  // Collect the values and masks in a keyset expression, this returns a vector
+  // where each element corresponds to the relevant tuple element in the select
+  // expression.
+  absl::optional<std::vector<std::pair<std::vector<bool>, std::vector<bool>>>>
+  CollectValuesAndMasks(const IR::Expression* keyset_expression) const;
+
+  // Report a compiler error associated with a location. This is a type-safe
+  // wrapper around `::error` from P4C that also handles some boilerplate for
+  // error reporting and keeping track of compiler errors.
+  template <typename... T>
+  void ReportErrorAt(Util::SourceInfo error_location,
```

Comment:
I agree about the `errorCount`, I am going to remove the `has_errors_` member and use `errorCount` instead.

Besides that, I use this function for two things:
1. To print the source fragment and position for each error message.
2. To output the arguments using `operator<<` so that I will not do any non-type-safe things with the format string.